### PR TITLE
Run lint checks on PRs, posting any warnings as comments directly on the files

### PR DESCRIPTION
**Pull request by jonsimantov**
This script performs a number of steps:
1. Get the PR's diff to find the list of affected files and lines in the PR.
2. Run cpplint on all files in the PR. For each lint warning, check if it falls within the range of lines affected by the diff. Omit that warning if it doesn't fall in the affected lines.
3. Delete any prior lint warning comments posted by previous runs (this takes a moment as they must be deleted one by one, due to GitHub API limitations)
4. Post all the lint warnings that fall within the range of the PR's diff (this is done all at once), on the PR's most recent commit.

It will ignore any lint warnings that occur outside the PR's diff, and it will only run cpplint on files that are added/modified in the PR in the first place. There is also an exclusion list to exclude sets of files (via regex).

Any other errors from cpplint (e.g. "I don't know how to parse this file") are simply ignored.

Also, the whole process will be skipped if the PR has the 'no-lint' label.

### CODE REVIEW

File: scripts/gha/lint_commenter.py [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+    for exclude_regex in EXCLUDE_PATH_REGEX:
+      if re.search(exclude_regex, pr_patch_file.path):
+        excluded = True
+        break
+    if excluded: continue
+    file_list.append(pr_patch_file.path)
+    pr_file_position = 1
+    line_reference[pr_patch_file.path] = {}
+    for hunk in pr_patch_file:
+      pr_file_position = pr_file_position + 1
```
Comment (DellaBitta) [inline scripts/gha/lint_commenter.py:144]: isn't this a problem that the pr_file_position starts at 2?

File: scripts/gha/lint_commenter.py [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+        ('--filter=%s' % CPPLINT_FILTER),
+        ('--repository=..')
+    ] + file_list, stderr=subprocess.STDOUT).decode('utf-8').split('\n')
+  except subprocess.CalledProcessError as e:
+    # Nothing to do if there is an exception.
+    lint_results=e.output.decode('utf-8').split('\n')
+
+  comments = []
+  for line in lint_results:
+    m = re.match(r'([^:]+):([0-9]+): *(.*[^ ]) +\[[^]]+\] \[\d+\]$', line)
```
Comment (DellaBitta) [inline scripts/gha/lint_commenter.py]: probably worth a comment as to what this is matching.

File: scripts/gha/lint_commenter.py [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+from unidiff import PatchSet
+import urllib.parse
+
+# Put any lint warnings you want to fully ignore into this list.
+IGNORE_LINT_WARNINGS = [
+    'build/header_guard',
+    'readability/casting',
+    'whitespace/line_length'
+]
+# Exclude files within the following paths (specified as regexes)
```
[the remark this answers is not in the document]
Comment (jonsimantov) [reply scripts/gha/lint_commenter.py:45]: We can add more as we go, that's the first batch I saw that comes from another repo and so had a lot of lint issues.

File: scripts/gha/lint_commenter.py (hunk 1 above)
Comment (jonsimantov) [reply scripts/gha/lint_commenter.py:144]: Technically no, according to https://docs.github.com/en/rest/reference/pulls#create-a-review-for-a-pull-request, the @@ at the start of the diff is position 1, the next line is position 2, etc.

File: scripts/gha/lint_commenter.py (hunk 2 above)
Comment (jonsimantov) [reply scripts/gha/lint_commenter.py]: Done.

File: scripts/gha/lint_commenter.py [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+
+  if args.verbose:
+    print('Commit sha:', commit_sha)
+
+  # Get the diff for the pull request.
+  # GET /repos/{owner}/{repo}/pulls/{pull_number}
+  request_url = 'https://api.github.com/repos/%s/%s/pulls/%s' % (repo_owner, repo_name, args.pr_number)
+  header = 'Accept: application/vnd.github.VERSION.diff'
+  if args.verbose:
+    print('request_url: %s' % request_url)
```
[the remark this answers is not in the document]
Comment (jonsimantov) [reply scripts/gha/lint_commenter.py:114]: Made more consistent.

File: scripts/gha/lint_commenter.py [hunk 5]
```diff
@@ -0,0 +1,280 @@
...
+      if label['name'] == LABEL_TO_SKIP_LINT:
+        skip_lint = True
+        break
+
+  if skip_lint:
+    print('PR #%s has "%s" label, skipping lint checks' % (args.pr_number, LABEL_TO_SKIP_LINT))
+    exit(0)
+
+  if args.verbose:
+    print('Commit sha:', commit_sha)
```
[the remark this answers is not in the document]
Comment (jonsimantov) [reply scripts/gha/lint_commenter.py]: Done. Good idea.

File: scripts/gha/lint_commenter.py (hunk 1 above)
Comment (jonsimantov) [reply scripts/gha/lint_commenter.py:146]: Oh wait you're right! I misread it. I'll fix it.

File: admob/src/android/banner_view_internal_android.h [hunk 6]
```diff
@@ -48,6 +48,7 @@ namespace internal {
 
 class BannerViewInternalAndroid : public BannerViewInternal {
  public:
+  // The following line should trigger a lint warning.
   BannerViewInternalAndroid(BannerView* base);
```
Comment (jonsimantov) [inline admob/src/android/banner_view_internal_android.h:52]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="admob/src/android/banner_view_internal_android.h:52:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/bundle_test.cc [hunk 7]
```diff
@@ -1,5 +1,6 @@
 // Copyright 2021 Google LLC
 
+// The following line should cause a lint warning.
 #include <future>
```
Comment (jonsimantov) [inline firestore/integration_test_internal/src/bundle_test.cc:4]: ⚠️ Lint warning: `<future> is an unapproved C++11 header.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/bundle_test.cc:4:  <future> is an unapproved C++11 header.  [build/c++11] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc [hunk 8]
```diff
@@ -1,6 +1,7 @@
 // Copyright 2021 Google LLC
 
 #include "firestore_integration_test.h"
```
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:3]: ⚠️ Lint warning: `Include the directory when naming .h files`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:3:  Include the directory when naming .h files  [build/include_subdir] [4]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc [hunk 9]
```diff
@@ -260,14 +261,10 @@ std::vector<std::string> FirestoreIntegrationTest::QuerySnapshotToIds(
   return result;
 }
 
-std::vector<MapFieldValue> FirestoreIntegrationTest::QuerySnapshotToValues(
-    const QuerySnapshot& snapshot) const {
-  std::vector<MapFieldValue> result;
```
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:265]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:266:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc [hunk 10]
```diff
@@ -306,6 +303,20 @@ bool FirestoreIntegrationTest::FailIfUnsuccessful(const char* operation,
   }
 }
 
+std::vector<MapFieldValue> FirestoreIntegrationTest::QuerySnapshotToValues(
+    const QuerySnapshot& snapshot) const {
+  std::vector<MapFieldValue> result;
+  for (const DocumentSnapshot& doc : snapshot.documents()) {
```
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:309]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:317:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: admob/src/android/banner_view_internal_android.h (hunk 6 above)
Comment (jonsimantov) [inline admob/src/android/banner_view_internal_android.h:52]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="admob/src/android/banner_view_internal_android.h:52:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/bundle_test.cc (hunk 7 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/bundle_test.cc:4]: ⚠️ Lint warning: `<future> is an unapproved C++11 header.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/bundle_test.cc:4:  <future> is an unapproved C++11 header.  [build/c++11] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 8 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:3]: ⚠️ Lint warning: `Include the directory when naming .h files`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:3:  Include the directory when naming .h files  [build/include_subdir] [4]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc [hunk 11]
```diff
@@ -260,14 +261,10 @@ std::vector<std::string> FirestoreIntegrationTest::QuerySnapshotToIds(
...
-    const QuerySnapshot& snapshot) const {
-  std::vector<MapFieldValue> result;
-  for (const DocumentSnapshot& doc : snapshot.documents()) {
-    result.push_back(doc.GetData());
-  }
-  return result;
-}
+class AnotherTemporaryLintErrorClass {
+  // The below line should give a lint error.
+  AnotherTemporaryLintErrorClass(int t) {}
```
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:266]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:266:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc [hunk 12]
```diff
@@ -306,6 +303,20 @@ bool FirestoreIntegrationTest::FailIfUnsuccessful(const char* operation,
...
+  std::vector<MapFieldValue> result;
+  for (const DocumentSnapshot& doc : snapshot.documents()) {
+    result.push_back(doc.GetData());
+  }
+  return result;
+}
+
+class TemporaryLintErrorClass {
+  // The below line should give a lint error.
+  TemporaryLintErrorClass(int t) {}
```
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:317]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:317:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: admob/src/android/banner_view_internal_android.h (hunk 6 above)
Comment (jonsimantov) [inline admob/src/android/banner_view_internal_android.h:52]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="admob/src/android/banner_view_internal_android.h:52:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/bundle_test.cc (hunk 7 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/bundle_test.cc:4]: ⚠️ Lint warning: `<future> is an unapproved C++11 header.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/bundle_test.cc:4:  <future> is an unapproved C++11 header.  [build/c++11] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 8 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:3]: ⚠️ Lint warning: `Include the directory when naming .h files`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:3:  Include the directory when naming .h files  [build/include_subdir] [4]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 11 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:266]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:266:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 12 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:317]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:317:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: admob/src/android/banner_view_internal_android.h (hunk 6 above)
Comment (jonsimantov) [inline admob/src/android/banner_view_internal_android.h:52]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="admob/src/android/banner_view_internal_android.h:52:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/bundle_test.cc (hunk 7 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/bundle_test.cc:4]: ⚠️ Lint warning: `<future> is an unapproved C++11 header.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/bundle_test.cc:4:  <future> is an unapproved C++11 header.  [build/c++11] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 8 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:3]: ⚠️ Lint warning: `Include the directory when naming .h files`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:3:  Include the directory when naming .h files  [build/include_subdir] [4]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 11 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:266]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:266:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 12 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:317]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:317:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: admob/src/android/banner_view_internal_android.h (hunk 6 above)
Comment (jonsimantov) [inline admob/src/android/banner_view_internal_android.h:52]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="admob/src/android/banner_view_internal_android.h:52:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/bundle_test.cc (hunk 7 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/bundle_test.cc:4]: ⚠️ Lint warning: `<future> is an unapproved C++11 header.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/bundle_test.cc:4:  <future> is an unapproved C++11 header.  [build/c++11] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 8 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:3]: ⚠️ Lint warning: `Include the directory when naming .h files`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:3:  Include the directory when naming .h files  [build/include_subdir] [4]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 11 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:266]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:266:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 12 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:317]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:317:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: admob/src/android/banner_view_internal_android.h (hunk 6 above)
Comment (jonsimantov) [inline admob/src/android/banner_view_internal_android.h:52]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="admob/src/android/banner_view_internal_android.h:52:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/bundle_test.cc (hunk 7 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/bundle_test.cc:4]: ⚠️ Lint warning: `<future> is an unapproved C++11 header.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/bundle_test.cc:4:  <future> is an unapproved C++11 header.  [build/c++11] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 8 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:3]: ⚠️ Lint warning: `Include the directory when naming .h files`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:3:  Include the directory when naming .h files  [build/include_subdir] [4]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 11 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:266]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:266:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: firestore/integration_test_internal/src/firestore_integration_test.cc (hunk 12 above)
Comment (jonsimantov) [inline firestore/integration_test_internal/src/firestore_integration_test.cc:317]: ⚠️ Lint warning: `Single-parameter constructors should be marked explicit.`<hidden value="cpplint-file-comment"></hidden><hidden value="firestore/integration_test_internal/src/firestore_integration_test.cc:317:  Single-parameter constructors should be marked explicit.  [runtime/explicit] [5]"></hidden>

File: scripts/gha/lint_commenter.py (hunk 1 above)
Comment (jonsimantov) [reply scripts/gha/lint_commenter.py]: I decided to give up on calculating the "position" (when I tried a diff with lines removed it got way more complicated) and am now just using "line", it's much simpler. I only need to get a list of the lines touched, not track each one's position in the diff.

**Review by DellaBitta — APPROVED** [review]
Are you going to remove those `// The below line should give a lint error` comments?

**Comment by jonsimantov** [comment]
> Are you going to remove those `// The below line should give a lint error` comments?

Yeah, I'll undo all of those lint warnings I added, it was just to demonstrate how it works. I'll need another approval after. :)

**Review by DellaBitta — APPROVED** [review]

**Pull request merged**